ts/nni_manager/common/trial_keeper/keeper.ts

```diff
@@ -0,0 +1,154 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+/**
+ *  The helper class to implement a "reuse" training service.
+ *
+ *  TrialKeeper has a very similar interface to TrainingSerivceV3.
+ *  In fact the local training service is a thin wrapper over TrialKeeper.
+ **/
+
+import { EventEmitter } from 'events';
+import fs from 'fs/promises';
+import path from 'path';
+
+import type { Command } from 'common/command_channel/interface';
+import { HttpChannelServer } from 'common/command_channel/http';
+import globals from 'common/globals';
+import { Logger, getLogger } from 'common/log';
+import { TrialProcess, TrialProcessOptions } from './process';
+import { GpuSchedulerClient } from './gpu_scheduler_client';
+
+export declare namespace TrialKeeper {
+    export interface TrialOptions {
+        id: string;
+        command: string;
+        codeDirectoryName: string;
+        sequenceId?: number;
+        gpuNumber?: number;
+        gpuRestrictions?: GpuRestrictions;
+    }
+
+    export interface GpuRestrictions {
+        onlyUseIndices?: number[];
+        rejectActive?: boolean;
+        rejectComputeActive?: boolean;
+    }
+}
+
+export class TrialKeeper {
+    private envId: string;
+    private channels: HttpChannelServer;
+    private dirs: Map<string, string> = new Map();
+    private emitter: EventEmitter = new EventEmitter();
+    private gpuScheduler: GpuSchedulerClient;
+    private log: Logger;
+    private platform: string;
+    private trials: Map<string, TrialProcess> = new Map();
+
+    constructor(environmentId: string, platform: string, enableGpuScheduler: boolean) {
+        this.envId = environmentId;
+        this.platform = platform;
+        this.log = getLogger(`TrialKeeper.${environmentId}`);
+
+        this.gpuScheduler = new GpuSchedulerClient(enableGpuScheduler);
+
+        this.channels = new HttpChannelServer(this.envId, `/env/${this.envId}`);
```

Comment:
When the experiment keeps running for a long time, is it possible that I get a large number of registered (and useless) HTTP servers?

Reply:
Each trial keeper is running inside an environment (or inside a worker in distributed trial's term). In NNI manager process there will be only one trial keeper (if local training service is enabled).

Reply:
I'm not sure I've fully understood. We'll see when remote is ready.

Reply:
what is this channel used for? only control plane? or also serves as data plane (i.e., metrics)?

Reply:
> what is this channel used for? only control plane? or also serves as data plane (i.e., metrics)?

Trial to manager:
- request parameter
- metrics
Manager to trial:
- parameter